src/PostgREST/App.hs

```diff
@@ -161,438 +113,47 @@ postgrestResponse
-  else
-    return response
-
-shouldCount :: Maybe PreferCount -> Bool
-shouldCount preferCount =
-  preferCount == Just ExactCount || preferCount == Just EstimatedCount
-
-returnsScalar :: ApiRequest.Target -> Bool
-returnsScalar (TargetProc proc _) = Proc.procReturnsScalar proc
-returnsScalar _                   = False
-
-readRequest :: Monad m => QualifiedIdentifier -> RequestContext -> Handler m ReadRequest
-readRequest QualifiedIdentifier{..} (RequestContext AppConfig{..} dbStructure apiRequest) =
-  liftEither $
-    ReqBuilder.readRequest qiSchema qiName configDbMaxRows
-      (dbRelationships dbStructure)
-      apiRequest
-
-contentTypeHeaders :: RequestContext -> [HTTP.Header]
-contentTypeHeaders RequestContext{..} =
-  ContentType.toHeader (iAcceptContentType ctxApiRequest) : maybeToList (profileHeader ctxApiRequest)
-
--- | If raw(binary) output is requested, check that ContentType is one of the
--- admitted rawContentTypes and that`?select=...` contains only one field other
--- than `*`
-binaryField :: Monad m => RequestContext -> ReadRequest -> Handler m (Maybe FieldName)
-binaryField RequestContext{..} readReq
-  | returnsScalar (iTarget ctxApiRequest) && iAcceptContentType ctxApiRequest `elem` rawContentTypes ctxConfig =
-      return $ Just "pgrst_scalar"
-  | iAcceptContentType ctxApiRequest `elem` rawContentTypes ctxConfig =
-      let
-        fldNames = fstFieldNames readReq
-        fieldName = headMay fldNames
-      in
-      if length fldNames == 1 && fieldName /= Just "*" then
-        return fieldName
-      else
-        throwError $ Error.BinaryFieldError (iAcceptContentType ctxApiRequest)
-  | otherwise =
-      return Nothing
-
-rawContentTypes :: AppConfig -> [ContentType]
-rawContentTypes AppConfig{..} =
-  (ContentType.decodeContentType <$> configRawMediaTypes) `union` [CTOctetStream, CTTextPlain]
-
-profileHeader :: ApiRequest -> Maybe HTTP.Header
-profileHeader ApiRequest{..} =
-  (,) "Content-Profile" <$> (toS <$> iProfile)
-
-splitKeyValue :: ByteString -> (ByteString, ByteString)
-splitKeyValue kv =
-  (k, BS8.tail v)
-  where
-    (k, v) = BS8.break (== '=') kv
+handleRequest :: Request -> DbHandler Wai.Response
+handleRequest req =
+  case req of
+    ReadRequest requestInfo ->
+      Response.readResponse <$> Query.readQuery requestInfo
+    CreateRequest requestInfo ->
+      Response.createResponse <$> Query.createQuery requestInfo
+    UpdateRequest requestInfo ->
+      Response.updateResponse <$> Query.updateQuery requestInfo
+    SingleUpsertRequest requestInfo ->
+      Response.singleUpsertResponse <$> Query.singleUpsertQuery requestInfo
+    DeleteRequest requestInfo ->
+      Response.deleteResponse <$> Query.deleteQuery requestInfo
+    InfoRequest dbStructure _ QualifiedIdentifier{..} ->
+      case find tableMatches $ dbTables dbStructure of
+        Just table -> return $ Response.infoResponse hasPK table
+        Nothing    -> throwError Error.NotFound
```

Comment:
In the discussion in #1839 we realized that we need to parse the request (arguments in querystring etc.) to properly handle the `OPTIONS` request.

Does it make sense to change the interface here right now already, so that `requestInfo` is also passed into `Response.infoResponse`?

Reply:
There actually are different `requestInfo` types depending on the request type. This is explicit in `Request` but is not visible in `App`, so I've changed the naming of the `requestInfo` variables to indicate this.

We already pass and ignore (`_`) ApiRequest to the `infoResponse` dispatch, which would be easy to pass through. But it will be best (and easy) to create a custom `requestInfo` for that case then.